src/librustc_typeck/check/mod.rs

```diff
@@ -5302,6 +5302,56 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         Some(original_span.with_lo(original_span.hi() - BytePos(1)))
     }
 
+    // Rewrite `SelfCtor` to `StructCtor`
+    pub fn rewrite_self(&self, def : Def, span : Span) -> (Def, DefId, Ty<'tcx>) {
+        let tcx = self.tcx;
+        match def {
```

Comment:
I know this was present in the original code too, but could you change this `match` to `if let` (we really just care whether we have `Def::SelfCtor` or not).

Reply:
Sure. Will also rename `rewrite_self` to the more accurate `rewrite_self_ctor`.